### PR TITLE
Enhancement to Play option that synchronizes image with sound.

### PR DESCRIPTION
**Pull request by GeoffZC**
This commit implements a synchronization method in DisplayImage class that is called by the Music class in the playTune method. Activated when play is selected in image display panel dropdown.

### CODE REVIEW

**Comment by jody** [comment]
Does this address some specific issue?  (Presumably one that is [_In progress_](https://github.com/MetroCS/imagelab/projects/1))  If so, please reference the issue and include its # .

**Comment by GeoffZC** [comment]
Yes, this pull request addresses Issue #2

**Comment by GeoffZC** [comment]
Do you want someone other then myself to explain the changes? Also do you want unit tests or just a description of how to achieve the correct results by running and using the program?

**Comment by GeoffZC** [comment]
git-rebase
If I do git rebase --interactive HEAD~7, there is only one commit made by me and several made by you @jody . If I do git -rebase --interactive HEAD there appears to be only one commit.

**Comment by jody** [comment]
This PR shows 2 commits (https://github.com/MetroCS/imagelab/pull/28/commits).

**Comment by jody** [comment]
You, as author, should provide explanation of the changes included in this PR.  The goal is primarily for the Reviewer(s) to be able to understand and better decide whether the PR should be merged.

It's also incumbent upon the author to demonstrate that the code (a) correctly addresses the associated issue, (b) is understandable and maintainable, and (c) does not regress the code, i.e., there are no new bugs or undesired behaviors introduced.

**Comment by jody** [comment]
The assumption is that your team has reviewed these changes prior to you issuing the PR.  One or more members of the team can each add a comment volunteering to do a review.  I can then assign them as reviewers directly to this PR and the PR can be added to the project board.

**Comment by GeoffZC** [comment]
I see those two commits, however I'm confused because the second one "Merge Pull Request #2 from GeoffZC/synchronize" seems to have been created in GitHub rather then in the terminal with Git. This same commit also seems to reference the initial commit with the comment "Committing changes for synchronization". Maybe I did something wrong when I was committing or there is a way to squash within GitHub? Is there a way to search for the commit using the index ID (ex. 682455bfa75c0a22ec8e5bf499fb83f3bde4ad4c)? 

The team has already reviewed this code prior to the PR. I will go ahead and add commentary on the changes.

**Comment by jody** [comment]
I'm not sure what's up with the multiple commits.  The commit messages aren't particularly clear.  I might guess that there's a commit associated with your local repo reflected in the push to your personal fork. That would then be accommodated by the PR from your personal fork to your organization's fork.  Not sure why it's showing as two commits.

File: imagelab/DisplayImage.java [hunk 1]
```diff
@@ -35,6 +36,11 @@
     private static int xspot = XMAX;
     /** y-coordinate for next window */
     private static int yspot = YMAX;
+    
+    /**Image Producer to set img*/
+    private MemoryImageSource source;
+    /**Used to get STD_DURATION*/
+    private Music music;
```
Comment (jody) [inline imagelab/DisplayImage.java:43]: If the only intent is to get STD_DURATION, why is instantiation needed?
Where is this used in the code?

File: imagelab/ImgProvider.java [hunk 2]
```diff
@@ -26,6 +26,7 @@
     int             pixheight;
     /** Image width in pixels. */
     int             pixwidth;
+    static Thread playThread;
```
Comment (jody) [inline imagelab/ImgProvider.java:29]: What is the intent of this class field?  (At the very least, this requires a Javadoc comment.)

File: imagelab/ImgProvider.java [hunk 3]
```diff
@@ -397,18 +400,20 @@ void save() {
...
      */
-    public void play() {
+      public void play() {
+        if(playThread!=null) playThread.stop();
+        playThread = new Thread(() -> {
         short[][] red = getRed();     // Red plane
         short[][] green = getGreen(); // Green plane
         short[][] blue = getBlue();   // Blue plane
-        short[][] bw = getBWImage();  // Black & white image
+        short[][] bw = getRed();  // Black & white image
```
Comment (jody) [inline imagelab/ImgProvider.java:409]: Why was this change necessary?  Why should "bw" contain "red" information?

File: imagelab/ImgProvider.java [hunk 4]
```diff
@@ -490,5 +497,25 @@ public void showPixNew(String name) {
...
+        return pixwidth;
+    }
+    
+    public int getPixHeight(){
+        return pixheight;
+    }
+    
+    public DisplayImage getDis(){
+        return dis;
+    }
```
Comment (jody) [inline imagelab/ImgProvider.java:519]: This entire section of the public API requires documentation.  At the very least, there needs to be Javadoc comments for the published API.

File: sound/Music.java [hunk 5]
```diff
@@ -201,6 +209,13 @@ public void playTune(Tune tune) {
         Iterator<Chord> tuneIt = tune.iterator();
         int line = 0;
         while (tuneIt.hasNext()) {
+            for(int i = 0; i<newRow; i++){
+                row[i] = pixels[i];
```
Comment (jody) [inline sound/Music.java:213]: Why is the variable row needed?  That is, why can't the variable pixels be used directly?

File: sound/Music.java [hunk 6]
```diff
@@ -201,6 +209,13 @@ public void playTune(Tune tune) {
...
+            }
+            
+            //oldRow = newRow-1;
+            improvider.getDis().synchronize(width, height, row);
```
Comment (jody) [inline sound/Music.java:217]: Why is the variable row needed?  That is, why can't the variable pixels be used directly?

File: sound/Music.java [hunk 7]
```diff
@@ -192,7 +192,15 @@ public void playChord(Chord chord, int defaultDuration) {
...
      * @param tune the tune to play
      */
-    public void playTune(Tune tune) {
+    public void playTune(Tune tune, ImgProvider imp) {
+        int heightTracker = 0;
+        ImgProvider improvider = imp;
+        int[] pixels = improvider.getPix();
+        int height = improvider.getPixHeight();
+        int width = improvider.getPixWidth();
+        int oldRow = 0;
```
Comment (jody) [inline sound/Music.java:201]: What are the intents of oldRow and newRow?  How are they used?  The variable names themselves are insufficient to explain their meaning or usage.

**Comment by GeoffZC** [comment]
I will update the code for maintainability (including JavaDocs) and then have a teammate review it. I will also apply a checkstyle to this build.

File: imagelab/DisplayImage.java [hunk 8]
```diff
@@ -16,7 +17,7 @@
     /** The image of this frame. */
     Image img;
     /** The display panel of this frame. */
-    private DisPanel pane;
+    private DynaPanel pane;
```
Comment (GeoffZC) [inline imagelab/DisplayImage.java:20]: Using DynaPanel Class instead of DisPanel class so that menu items can still be selected while the music is playing and image is synchronizing.

File: imagelab/DisplayImage.java [hunk 9]
```diff
@@ -3,6 +3,7 @@
 import java.awt.*;
 import java.awt.image.*;
 import javax.swing.*;
+import sound.*;
```
Comment (GeoffZC) [inline imagelab/DisplayImage.java:6]: Will remove and re-commit this unnecessary package.

File: imagelab/DisplayImage.java [hunk 10]
```diff
@@ -35,6 +36,11 @@
     private static int xspot = XMAX;
     /** y-coordinate for next window */
     private static int yspot = YMAX;
+    
+    /**Image Producer to set img*/
+    private MemoryImageSource source;
```
Comment (GeoffZC) [inline imagelab/DisplayImage.java:41]: Will remove and re-commit this unnecessary instance variable.

File: imagelab/DisplayImage.java (hunk 1 above)
Comment (GeoffZC) [reply imagelab/DisplayImage.java:43]: I have removed this object instantiation as STD_DURATION  was only used as a preliminary attempt for image synchronization. The constant is no longer used or needed for synchronization.

File: imagelab/DisplayImage.java [hunk 11]
```diff
@@ -68,7 +74,7 @@ public DisplayImage(ImgProvider imp, String title, boolean slow) {
         int pixwidth  = imp.pixwidth;
         img = getToolkit().createImage(
                 new MemoryImageSource(pixwidth, pixheight, showpix, 0, pixwidth));
-        pane        = new DisPanel(img);
+        pane        = new DynaPanel(img);
```
Comment (GeoffZC) [inline imagelab/DisplayImage.java:77]: Initialization of new DynaPanel object using img object. The purpose here is to use the functionality of DynaPanel that allows for dropdown menus to be selected while music is playing and image is synchronizing.

File: imagelab/DisplayImage.java [hunk 12]
```diff
@@ -94,5 +100,17 @@ public DisplayImage(ImgProvider imp, String title, boolean slow) {
         setJMenuBar(myMenuBar);
         setVisible(true);
     }
+    
+    public void remove(){
```
Comment (GeoffZC) [inline imagelab/DisplayImage.java:104]: I have removed this remove method and will re-commit those changes. This was only used as a preliminary attempt to support the synchronization process but it is understood that this is no longer needed. A call to this method in the ImgProvider.java source has also been removed.

File: imagelab/DisplayImage.java [hunk 13]
```diff
@@ -94,5 +100,17 @@ public DisplayImage(ImgProvider imp, String title, boolean slow) {
...
+        getContentPane().remove(pane);
+    }   
+    
+     public void synchronize(int w, int h, int[] pixs){
```
Comment (GeoffZC) [inline imagelab/DisplayImage.java:108]: This method directly addresses issue #2. The method is supplied portions of pixels from the pix array that is filled by pixel grabber in ImgProvider.java. The pixel data is first passed to the playTune method from Music.java that is called inside the play method in ImgProvider.java when the play menu item is selected within the user interface. 
This synchronize method is called for each iteration of the tuneIt iterator inside the playTune method in Music.java. The amount of rows of pixels that are passed to this method are incremented starting at one and increasing by the same amount for each chord so that the image finishes rendering at the same time the music stops playing. 
I will leave more comments about this process in the playTune method in the Music.java class.

File: imagelab/DisplayImage.java [hunk 14]
```diff
@@ -94,5 +100,17 @@ public DisplayImage(ImgProvider imp, String title, boolean slow) {
...
+        source = new MemoryImageSource(w, h, pixs, 0, w);
```
Comment (GeoffZC) [inline imagelab/DisplayImage.java:109]: I have removed the use of this instance variable and just used a local variable. This will be provided in a new commit and PR.

File: imagelab/DynaPanel.java [hunk 15]
```diff
@@ -57,7 +57,7 @@ public DynaPanel() {
     public DynaPanel(Image img) {
         super(true);
         panelImage = img;
-        System.out.println("DynaPanel: panelImage = " + panelImage);
+        //System.out.println("DynaPanel: panelImage = " + panelImage);
```
Comment (GeoffZC) [inline imagelab/DynaPanel.java:60]: Removed output for "cleaner" use of the program.

File: imagelab/DynaPanel.java [hunk 16]
```diff
@@ -69,15 +69,17 @@ public DynaPanel(ImgProvider imp) {
         int width  = imp.getWidth();
         int height = imp.getHeight();
         panelImage = imp.getImage();
-        System.out.println("DynaPanel: panelImage = " + panelImage);
+        System.out.println("Using DynaPanel");
```
Comment (GeoffZC) [inline imagelab/DynaPanel.java:72]: Adjusting output to match new use of DynaPanel.

File: imagelab/DynaPanel.java [hunk 17]
```diff
@@ -69,15 +69,17 @@ public DynaPanel(ImgProvider imp) {
...
+        //System.out.println("DynaPanel: panelImage = " + panelImage);
     }
 
     /**
      * Draw a given image onto the canvas.
      * @param image the image object to be drawn on the canvas
      */
     public void newImage(Image image) {
-        System.out.println("DynaPanel:draw(i): image = " + image);
+        System.out.println("Using DynaPanel");
```
Comment (GeoffZC) [inline imagelab/DynaPanel.java:81]: I will remove this output as it was only used in testing to confirm use of new class.

File: imagelab/ImgProvider.java (hunk 2 above)
Comment (GeoffZC) [reply imagelab/ImgProvider.java:29]: I have added Javadoc to this and will resubmit PR.

File: imagelab/ImgProvider.java (hunk 3 above)
Comment (GeoffZC) [reply imagelab/ImgProvider.java:409]: Without this change, the image that was rendered during synchronization was in black and white since the toBW method converts the pix array from color to gray scale. I work on a fix to avoid this change.

File: imagelab/ImgProvider.java [hunk 18]
```diff
@@ -397,18 +400,20 @@ void save() {
...
-        short[][] bw = getBWImage();  // Black & white image
+        short[][] bw = getRed();  // Black & white image
         short[][] alpha = getAlpha(); // Alpha channel
         short[][] hue;
         short[][] saturation;
         short[][] brightness;
    
-        int height = bw.length;
-        int width  = bw[0].length;
+        int height = red.length;
```
Comment (GeoffZC) [inline imagelab/ImgProvider.java:415]: Change allowed synchronized image to be displayed in color.

File: imagelab/ImgProvider.java [hunk 19]
```diff
@@ -397,18 +400,20 @@ void save() {
...
+        int width  = red[0].length;
```
Comment (GeoffZC) [inline imagelab/ImgProvider.java:416]: Change allowed synchronized image to be displayed in color.

File: imagelab/ImgProvider.java [hunk 20]
```diff
@@ -465,7 +470,9 @@ public void play() {
         }//for row
         int[] instruments = {Note.Vibes, Note.Pizzacatto, Note.MelodicTom};
         Music m = new Music(3, instruments);
-        m.playTune(tune);
+        m.playTune(tune, this);
+        });
+        playThread.start();
```
Comment (GeoffZC) [inline imagelab/ImgProvider.java:475]: Starts the thread with the play method in it given a playThread is not already running play.

File: imagelab/ImgProvider.java [hunk 21]
```diff
@@ -490,5 +497,25 @@ public void showPixNew(String name) {
         dImage1.changeImage(this,"Second Pass");
         System.out.println("ImgProvider:showSlow: Second Pass");
     }
+    
+        public void removeContainer(){
```
Comment (GeoffZC) [inline imagelab/ImgProvider.java:501]: I have removed this method as it is not used. The change will be in the next PR.

File: imagelab/ImgProvider.java [hunk 22]
```diff
@@ -490,5 +497,25 @@ public void showPixNew(String name) {
...
+        dis.remove();
+    }
+
```
Comment (GeoffZC) [inline imagelab/ImgProvider.java:504]: Each of the below getter methods are used by the Music.java method playTune() in order to get information about the pix array and DisplayImage object.

File: imagelab/ImgProvider.java (hunk 4 above)
Comment (GeoffZC) [reply imagelab/ImgProvider.java]: I have added JavaDocs to these getters. They are available in the latest PR.
